### ISSUE TITLE
Follow on a non-Actor object

### ISSUE THREAD

**Issue by jaywink** [w3c-social/activitypump#343]
There is some confusion in the ActivityPub specification on what a `Follow` can be made towards.

The C2S part says "The Follow activity is used to subscribe to the activities of another actor."

https://www.w3.org/TR/activitypub/#follow-activity-outbox

This seems wrong since IMHO it should be a perfectly fine use case to `Follow { Note }` for example which correlates directly to the Diaspora protocol `Participation` (https://diaspora.github.io/diaspora_federation/entities/participation.html) which is an important mechanism to subscribe for follow-ups or updates.

This discussion came up in the ForgeFed specification building on top of ActivityPub and thinking whether an `Issue` also needs to be an actor.

Forum link: https://talk.feneas.org/t/following-non-actors/159

**Comment by nightpool** [comment]
I think it's perfectly fine to want to use the as2 verb `Follow` for such cases, but it's important to understand that it's currently outside of the scope of the ActivityPub spec and would need an extension (not least of which to the vocabulary, as we already discussed on IRC, so you know which actor you're expecting the Accept from)

While the s2s spec never explicitly comes out and *says* that the correct domain for the object property of a Follow is `Actor`, it implies it in many places, for example the SHOULD requirement quoted below:

> the server SHOULD add the actor to the object actor's Followers Collection. 

Obviously, non-actors do not having Follower collections and cannot fulfill this requirement. More generally, many existing systems, such as mastodon & pleroma, are only set up for following Actors and you have to think about what compatibility with such systems means.

In lieu of a dedicated UI action for following non-Actor objects, how do you expect to be compatible with such systems? (Do you expect to be compatible with such systems?) What does the user experience of opening a ticket / replying to a ticket / subscribing to a ticket look like on microblogging systems? These seems like useful use-cases to dig into—they'll help clarify whether you should use a backwards incompatible extension or work to fit issues into the Actor model.

**Comment by jaywink** [comment]
> but it's important to understand that it's currently outside of the scope of the ActivityPub spec and would need an extension

I'm finding it hard to believe that it was outscoped on purpose. It feels like the spec was mainly thinking of the "people will generally only follow actors" case but failed to consider other cases. It feels like the simplest solution to following an object shouldn't be taken just because this consideration wasn't done. 

> (not least of which to the vocabulary, as we already discussed on IRC, so you know which actor you're expecting the Accept from)

I'm not sure I understand. I think the chat discussion was related to an `Issue` which would not (under current plans by fr33domlover) be managed by the person who the issue is `attributedTo`. I don't personally believe there is a need for a managedBy attribute for Issues. This is offtopic really for this issue tracker and this issue here as it's purely a ForgeFed consideration. For a normal `Note` for example, you would know the actor from the `attributedTo`.

Also note that responding with `Accept` to a `Follow` is not mandated by ActivityPub, it's a SHOULD.

> Obviously, non-actors do not having Follower collections and cannot fulfill this requirement. 

The spec doesn't forbid a Note to have a follower collection either I believe? Also, adding the follower to the `followers` collection is a SHOULD, not a MUST.

> More generally, many existing systems, such as mastodon & pleroma, are only set up for following Actors and you have to think about what compatibility with such systems means.

Not every platform can be compatible in every way because not all platforms support all features. Besides, would it be impossible for Mastodon to support subscribing to replies of a Note for example? This would improve the reliability of getting all replies on a thread for example to the level that it is outside of microblogging environments. But whether or not Mastodon would want to do that, I don't think the specifics of how ActivityPub can be used should be considered based on the features of particular platforms. Mastodon and Pleroma are both microblogging platforms which is only a subset of social networking. When we look at "macroblogging" (like Diaspora, Friendica, etc), collecting all replies under the main post becomes much more important, for example. This is why the Diaspora protocol has the `Participation` entity, which works wonderfully for this purpose. I would imagine a similar need is for a forge Issue, which is where this conversation started from.

**Comment by nightpool** [comment]
It feels like much of your argument is "this would be useful—we should have it". I agree. I just don't think a reasonable person, reading the ActivityPub spec, would come away with any expectation that anything except actors could Follow and be Followed. I don't think the spec's requirements or semantics around Following make sense for non-actors, and I think sitting down and thinking what a good set of semantics and requirements for following objects looks like is better then trying to shoe-horn in the ones for actors.

I also don't believe it was outscoped on purpose. Again, I think it's a useful and valid use case. I just don't think it's covered by the current spec, and we should write an extension to cover it

**Comment by trwnh** [comment]
I get the feeling ActivityPub was written with message-passing in mind, and there's nothing explicitly saying that objects can't pass messages between each other. But those objects don't have to correspond to users. So the "vanilla" option would be to just make those objects into actors, if you expect to have them be followed. However, you also have the option of establishing some service to announce only the relevant activities to you (like how Github sends you emails*), as long as you send the appropriate signal... but there just isn't anything already specified to indicate interest in an object, or in only a subset of activities**. Still, it should also be noted that the Activity Vocabulary describes the Follow activity as such:

>Indicates that the actor is "following" the object. Following is defined in the sense typically used within Social systems in which the actor is interested in any activity performed by or on the object.

So what's *really* needed is a standardized mechanism to indicate to an actor that you're only interested in *some* activities, not *all*. Otherwise, you're stuck with pull-based polling to check when a Collection has been updated. Or you can follow the actor and get spammed with every single activity, discarding all of the ones you're not interested in.

*This is really analogous to subscribing to a mailing list, but doing so over email. Typically you get added to a mailing list out-of-band, but you could also theoretically reply with a keyword like "subscribe", as SMS mailing lists often do. In Github's case, the mailing list management happens through your account settings.

**(There's "streams", but that's very poorly defined.)

**Comment by jaywink** [comment]
@nightpool absolutely, I didn't mean this as a "fix immediately" issue. More of a thing to keep an eye on when the next draft is iterated upon some day.

On the [Feneas forum](https://talk.feneas.org/t/following-non-actors/159), Hubzilla/Zot/Zap developer Mike Macgirvin indicated that Zap has support for this kind of thing and Friendica developer Michael Vogel indicated Friendica already does this with ActivityPub. Socialhome will also be supporting `Follow  { Note }` as an example.

**Comment by ap-socialhub** [comment]
This issue was moved to https://socialhub.activitypub.rocks/t/follow-on-a-non-actor-object/169

**Comment by ap-socialhub** [comment]
This issue was moved to https://socialhub.activitypub.rocks/t/follow-on-a-non-actor-object/195

**Comment by evanp** [comment]
So, an `Actor` is anything that can generate activities and that might be addressed in activities. As the spec says:

> ActivityPub actors are generally one of the [ActivityStreams Actor Types](https://www.w3.org/TR/activitystreams-vocabulary/#actor-types), but they don't have to be.

If you want to follow a `Note`, check if it has an `inbox` property. If so, send it a `Follow`. If you get back an `Accept`, it's followable.

If you want to make your `Note` objects followable, give them an `inbox`. If you get a `Follow`, `Accept` it. When things happen with the `Note`, have the `Note` object `Announce` them.

I'll add some more information on this to the ActivityPub primer. But it's really important that `Person` objects aren't the only actors in the fediverse! Big bummer if so!

**Comment by trwnh** [comment]
i think it should be done with inbox forwarding, not Announce?

**Comment by evanp** [comment]
@trwnh Inbox forwarding should be implemented with an `Announce` activity from the actor or server.

**Comment by trwnh** [comment]
that sounds very contentious -- `Announce` has an existing semantic meaning that is more akin to a post kind than a forwarding mechanism. it gets added to `shares` and would be rendered as its own activity in the stream.

as i understand it, the mechanism for inbox forwarding is to POST the same body exactly as you received it, to any inboxes discovered by dereferencing collections owned by you. the only complication is with verifying or authenticating such forwarded activities, which led to projects like Mastodon adopting LD Signatures, and FEPs like 8b32 "Object Integrity Proof" to be worked on.

**Comment by evanp** [comment]
So, we've come back to this topic after almost a year. Here are the options I see before us:

- Better documentation about following non-AS2-Actor-type ActivityPub actors. (The definitions here are different, so I'm calling them out in the discussion). I think this meets the needs of the question, although there's a lot of conceptual slippage in here.
- Other mechanisms besides ActivityPub actor structure for following an object of any kind. One that was mentioned is prospective-search-based Follow, so that I define some properties, and receive all new activities that meet those properties, modulo authorization. Another is a separate way of following non-AS2-Actor-type actors, if needed.

My sense is that we probably need to do at least the first, maybe the second. I'm going to self-assign this one, mark it as needs Primer page, and also needs FEP.

**Comment by trwnh** [comment]
https://w3id.org/fep/efda and https://socialhub.activitypub.rocks/t/fep-efda-followable-objects/5030/1 are now available for this issue.

TLDR:

- To make something followable, give it a `followers` collection.
- If there isn't an `inbox` directly on the resource, then check `attributedTo`. Recurse as needed, possibly max depth 1. Send the `Follow` to the discovered inbox.

This roughly mirrors the interaction flow for any other object. For example, if you encounter a `Note` and you want to `Like` it, you don't directly check for `Note.inbox`, you generally check for `Note.attributedTo.inbox` and send it there.

Tangential note: Perhaps this flow can be generalized such that "if you want to interact with an object and it is missing an inbox, you MAY recurse upward through `attributedTo`"? I know that in certain cases a resource might have an `inbox` directly on it, and this suggests that you can send your notification directly to that inbox. (This is especially the case with LDN, maybe?) So maybe we could use a general Primer page for "how to interact" or "where to send your activity". (EDIT: see #493 for this)

**Comment by evanp** [comment]
We added a primer page to give guidance on this issue.

https://www.w3.org/wiki/ActivityPub/Primer/Actors

Please review and we will close once it's got feedback.

**Comment by silverpill** [comment]
>An important thing to note is that actors are identified using [duck-typing](https://en.wikipedia.org/wiki/Duck_typing), that is, they are identified by whether they have a certain set of properties. There is no Actor type defined in ActivityPub.

@evanp While I agree that duck typing is the best solution, it doesn't automatically follow from ActivityPub requirements and is somewhat controversial. Also, it is not an established practice (most implementations recognize only actor types from Activity Vocabulary, and for a good reason, as duck-typing may have unintended consequences).

I think ideas of this kind should be developed as FEPs. [FEP-2277: ActivityPub core types](https://codeberg.org/fediverse/fep/src/branch/main/fep/2277/fep-2277.md) already exists, and describes how to duck-type an actor in greater detail. Consider referring to it.

>[Content type objects](https://www.w3.org/TR/activitystreams-vocabulary/#object-types) can be actors, just like actor-type objects.

The link leads to Activity Vocabulary, but it doesn't define content type objects. Did you mean simply "objects"? Or, if this means a new class of objects that is disjoint with actors, activities and collections, then again FEP-2277 might be helpful, because it defines disjoint classes.

>What can go in the outbox
>TBD, possibly inbox forwarding stuff 

No, ActivityPub says that outbox contains activities "published" and "produced" by an actor, and forwarding is done by servers (we talked about it in another issue: https://github.com/w3c/activitypub/issues/494#issuecomment-2676993576). Inclusion of activities from other actors in outbox is a violation of the spec and is a change that may break interoperability with conformant implementations.

**Comment by silverpill** [comment]
Regarding the original issue topic, I agree with other commenters who said that ActivityPub implies that only actors could be followed. [FEP-efda: Followable objects](https://codeberg.org/fediverse/fep/src/branch/main/fep/efda/fep-efda.md) proposes a new behavior for all `Follow` activities. I published another FEP which is compatible with ActivityPub on a conceptual level (only actors could be followed) and doesn't require `attributedTo` traversal (it introduces a new property with a specific purpose): [FEP-f06f: Object observers](https://codeberg.org/fediverse/fep/src/branch/main/fep/f06f/fep-f06f.md).

**Comment by evanp** [comment]
Hey, @silverpill . So, the text of the ActivityPub spec is very clear that actors don't have to be of any particular type. It only defines properties for an actor, not a type. That's what duck typing means.

I'm well aware that there are non-compliant implementations that don't handle this well. That's a bug in those implementations, and we need to fix them. The primer calls out this fact already.

One workaround is to multi-type with an actor type, like ["Note", "Application"], but the servers that don't handle non-"actor type" actors also don't handle multityping as well. :/

I agree that only ActivityPub actors can be followed. But an `Image` or a `Place` can be an ActivityPub actor, even if they don't have an AS2 "actor type".

It is really difficult to deal with the fact that we use "actor" in three different ways in AP/AS2 (the `actor` property, AS2 "actor" types, and ActivityPub actors) and they don't necessarily overlap!

I understand your concern with the link titled "content-type objects" which links to the "object and link types" in Activity Vocabulary. I'll change it to "non-actor types" instead. I think the overloading is already pretty rough, and explaining that some objects use "object types" is a little more than I wanted to put on the reader.

**Comment by evanp** [comment]
Given that this has been open for a while, that it is documented in the primer, and is based on core functionality in AP, I'm marking it pending closure.

**Comment by silverpill** [comment]
>I'm well aware that there are non-compliant implementations that don't handle this well. That's a bug in those implementations, and we need to fix them. The primer calls out this fact already.

That is not a bug. There are 3 ways to identify an actor with non-standard `type`, e.g. `Feed`:
- Define `Feed` as an actor type (just like we define `Person` to be an actor type).
- Multi-type with a standard type: `["Person", "Feed"]`
- Duck type it by `inbox` and `outbox`

As far as I can tell, all 3 are perfectly valid according to specifications.

**Comment by trwnh** [comment]
> Define Feed as an actor type (just like we define Person to be an actor type).

We do not currently define this, because the concept of an "actor" is not sufficiently defined in a disambiguated manner. The "bug" being referred to is expecting only certain "actor types" instead of being able to handle any type. ActivityPub explicitly calls out the possibility of AP actors being any type. Attempting to enforce an allowlist of "actor types" is what causes the issue with overly strict validation.

> Inclusion of activities from other actors in outbox is a violation of the spec and is a change that may break interoperability with conformant implementations.

This is exactly the kind of overly strict validation that prevents general use cases from being handled correctly. Per AP, an `outbox` is required to contain `Activity` items, and it is required to be reverse-chronological. Nothing is said about the `actor` property nor is it required to be a singular value shared across every single item. The `outbox` is basically a feed of whatever the AP actor "publishes". In the purest sense, whoever controls the HTTP resource can serve whatever links they want in their resource, and thus `outbox` is valid on any resource, just like how `inbox` is valid on any resource.

---

Given that the side effects of a `Follow` activity are defined in terms of manipulating the `followers` collection, I don't see why the definition of an "actor" is relevant here; it is more of a red herring than anything useful. As the sender of a `Follow` activity, you know your follow was successful when you get an `Accept Follow` back. At that point, you can assume you are in the `followers` collection, even if the `followers` collection is not exposed to you. Likewise, the contents of the `outbox` are not really relevant to what gets delivered to you because you are in `followers`; it's only relevant when you GET the outbox. This can be considered a form of "backfill" in the same way that browsing old RSS/Atom items works.

**Comment by silverpill** [comment]
Actor types are defined in Activity Streams vocabulary: `Person`, `Application`, etc. ActivityPub doesn't require applications to handle any type, it simply states that other types _might_ be used:

>ActivityPub actors are generally one of the ActivityStreams Actor Types, but they don't have to be. For example, a Profile object might be used as an actor, or a type from an ActivityStreams extension.

According to this, all 3 methods I listed earlier are valid. If there is some other text in the specification supporting the claim about duck typing being the only correct method, please cite it. Otherwise, this Primer page should be edited or removed, because it is misleading.

Also, none of the other suggested changes have been made yet.

**Comment by evanp** [comment]
> > I'm well aware that there are non-compliant implementations that don't handle this well. That's a bug in those implementations, and we need to fix them. The primer calls out this fact already.
> 
> That is not a bug. There are 3 ways to identify an actor with non-standard `type`, e.g. `Feed`:
> 
> * Define `Feed` as an actor type (just like we define `Person` to be an actor type).
> * Multi-type with a standard type: `["Person", "Feed"]`
> * Duck type it by `inbox` and `outbox`
> 
> As far as I can tell, all 3 are perfectly valid according to specifications.

ActivityPub actors and Activity Streams 2.0 "actor type" objects are distinct. An object with an AS2 actor type is not necessarily an AP actor.

For ActivityPub, there is only one way to mark an object as an ActivityPub actor, and that is by giving it an `inbox` and an `outbox`. An object without these properties, including objects that have types that are Activity Streams 2.0 "actor types", are not ActivityPub actors.

I realize this is confusing; I'm going to go over the primer page one more time to make this a little clearer.

**Comment by evanp** [comment]
We reviewed one more time, and there is ample discussion of how actors are typed in ActivityPub. I'm going to close this ticket.
